src/main/java/ch/njol/skript/expressions/ExprSets.java

```diff
@@ -39,152 +29,74 @@
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Supplier;
 
 @Name("Sets")
-@Description("Collection sets of items or blocks of a specific type or colours, useful for looping.")
+@Description("A list of all the values of a type; useful for looping.")
 @Examples({
-		"loop items of type ore and log:",
-				"\tblock contains loop-item",
-				"\tmessage \"Theres at least one %loop-item% in this block\"",
-		"drop all blocks at the player # drops one of every block at the player"
+	"loop all attribute types:",
+	"\tset loop-value attribute of player to 10",
+	"\tmessage \"Set attribute %loop-value% to 10!\""
 })
-@Since("<i>unknown</i> (before 1.4.2), INSERT VERSION (colors)")
-public class ExprSets extends SimpleExpression<Object> {
+@Since("INSERT VERSION")
+public class ExprSets<T> extends SimpleExpression<T> {
 
 	static {
 		Skript.registerExpression(ExprSets.class, Object.class, ExpressionType.COMBINED,
-				"[(all [[of] the]|the|every)] item(s|[ ]types)", "[(all [[of] the]|the)] items of type[s] %itemtypes%",
-				"[(all [[of] the]|the|every)] block(s|[ ]types)", "[(all [[of] the]|the)] blocks of type[s] %itemtypes%",
-				"([all [[of] the]] colo[u]rs|(the|every) colo[u]r)");
+			"[(all [[of] the]|the|every)] %*classinfo%");
 	}
 
 	@Nullable
-	private Expression<ItemType> types;
-	private int pattern = -1;
+	private Supplier<Iterator<T>> supplier;
+	private ClassInfo<T> classInfo;
 
-	@SuppressWarnings("unchecked")
 	@Override
-	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
-		if (vars.length > 0)
-			types = (Expression<ItemType>) vars[0];
-		pattern = matchedPattern;
-		if (types instanceof Literal) {
-			for (ItemType type : ((Literal<ItemType>) types).getAll())
-				type.setAll(true);
-		}
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
+			classInfo = ((Literal<ClassInfo<T>>) exprs[0]).getSingle();
+			supplier = classInfo.getSupplier();
+			if (supplier == null) {
+				Skript.error("You cannot get all values of type '" + classInfo.getName().getSingular() + "'");
+				return false;
+			}
 		return true;
 	}
 
-	private Object[] buffer = null;
-
 	@Override
-	protected Object[] get(Event event) {
-		if (buffer != null)
-			return buffer;
-		List<Object> elements = new ArrayList<>();
-		for (Object element : new IteratorIterable<>(iterator(event)))
-			elements.add(element);
-		if (types instanceof Literal)
-			return buffer = elements.toArray();
-		return elements.toArray();
+	@SuppressWarnings("unchecked")
+	protected T[] get(Event event) {
+		Iterator<T> iterator = supplier.get();
+		List<T> elements = new ArrayList<>();
+		iterator.forEachRemaining(elements::add);
+		return (T[]) elements.toArray(new Object[0]);
```

Comment:
```suggestion
		return (T[]) elements.toArray(new Object[elements.size()]);
```

Reply:
It's faster to not predefine a size if we aren't using it somewhere else, as seen here
![image](https://user-images.githubusercontent.com/98935832/190434620-b60f3091-9b6a-4694-abc2-6005a61cf4e9.png)

Reply:
I always use `toArray(Object[]::new)` which is even better.

Reply:
I think using 0 is fine. We can't do `list.toArray(Object[]::new)` as that's a Java 11 feature.

Reply:
Java handles that when building a Java 8 version with Gradle.

Reply:
`list.toArray(Object[]::new)` calls `list.toArray(new Object[0])` making it actually slower because it does an extra call.
Here's the source code for `list.toArray(Object[]::new)`
![image](https://user-images.githubusercontent.com/98935832/192119415-920eb1f5-fd9e-40b2-b28d-ad32f8018b09.png)